api/app/Http/Controllers/ApiController.php

```diff
@@ -0,0 +1,218 @@
+        if ($route = RouteHelper::getRoute($model)) {
+            $response->setContent(json_encode([$route]));
+        }
+        return $response;
+    }
+
+    /**
+     * Put an entity.
+     *
+     * @param  string   $id
+     * @param  Request  $request
+     * @return Response
+     */
+    public function putOne($id, Request $request)
+    {
+        try {
+            $model = $this->repository->find($id);
+        } catch (ModelNotFoundException $e) {
+            $model = $this->repository->getModel();
+        }
+        $model->fill($request->all());
+        $this->repository->save($model);
+
+        $response = $this->responder->created();
+        if ($route = RouteHelper::getRoute($model)) {
+            $response->setContent(json_encode([$route]));
+        }
+
+        return $response;
+    }
+
+    /**
+     * Put many entities.
+     *
+     * @param  Request  $request
+     * @return Response
+     */
+    public function putMany(Request $request)
+    {
+        $data = $request->data;
+        $idName = $this->repository->getModel()->getKeyName();
+        $ids = [];
+        foreach ($data as $piece) {
+            $ids[] = $piece[$idName];
+        }
+        $models = $this->repository->findMany($ids);
+
+        $putModels = [];
+        foreach ($data as $piece) {
+            $id = $piece[$idName];
+            if ($models->has($id)) {
+                $model = $models->get($id);
+            } else {
+                $model = $this->repository->getModel();
+            }
+            /** @var BaseModel $model */
+            $model->fill($piece);
+            $putModels[] = $model;
+        }
+
+        $models = $this->repository->saveMany($putModels);
+
+        $response = $this->responder->created();
+        $routes = [];
+        foreach ($models as $model) {
+            if ($route = RouteHelper::getRoute($model)) {
+                $routes[] = $route;
+            }
+        }
+        $response->setContent(json_encode($routes));
+
+        return $response;
+    }
+
+    /**
+     * Patch an entity.
+     *
+     * @param  string   $id
+     * @param  Request  $request
+     * @return Response
+     */
+    public function patchOne($id, Request $request)
+    {
+        $model = $this->repository->find($id);
+        $model->fill($request->all());
+        $this->repository->save($model);
+
+        return $this->responder->noContent();
+    }
+
+    /**
+     * Patch many entites.
+     *
+     * @param  Request  $request
+     * @return Response
+     */
+    public function patchMany(Request $request)
+    {
+        $data = $request->data;
+        $idName = $this->repository->getModel()->getKeyName();
+        $ids = [];
+        foreach ($data as $piece) {
+            $ids[] = $piece[$idName];
+        }
+        $models = $this->repository->findMany($ids);
+        if ($models->count() !== count($ids)) {
+            throw new \InvalidArgumentException('Not all entities were found');
```

Comment:
Yes actually on further reflections, it should be the same as a validation exception, see https://github.com/spira/spira/wiki/Exception-format#validation-of-multiple-entities 
For example:
```
{
    "message": "There was an issue with the validation of provided entities",
    "invalid": [
        null,
        null,
        {
            "entityId": [
                {
                    "type": "notFound",
                    "message": "Entity not found"
                }
            ]
        },
        null
    ]
}
```